networkx/algorithms/structuralholes.py

```diff
@@ -138,6 +138,55 @@ def redundancy(G, u, v, weight=None):
         )
         return 1 - r
 
+    try:
+        import numpy as np
+
+        has_numpy = True
+    except:
+        has_numpy = False
+
+    if nodes is None and has_numpy:
+        # In order to compute constraint of all nodes,
+        # algorithms based on sparse matrices can be much faster
+
+        # Obtain the adjacency matrix
+        P = nx.adjacency_matrix(G, weight=weight)
+
+        # Calculate mutual weights
+        mutual_weights1 = P + P.T
+        mutual_weights2 = mutual_weights1.copy()
+
+        # Mutual_weights1 = Normalize mutual weights by row sums
+        sum_mutual_weights = mutual_weights1.sum(axis=1)
+        val = np.asarray(
+            np.repeat(sum_mutual_weights, mutual_weights1.getnnz(axis=0))
+        ).flatten()
+        mutual_weights1.data = mutual_weights1.data / val
+
+        # Mutual_weights2 = Normalize mutual weights by row max
+        max_mutual_weights = mutual_weights2.max(axis=1).todense()
+        val = np.asarray(
+            np.repeat(max_mutual_weights, mutual_weights2.getnnz(axis=0))
+        ).flatten()
+        mutual_weights2.data = mutual_weights2.data / val
+
+        # Calculate effective sizes
+        n_nodes = len(G)
+        r = np.ones((n_nodes, n_nodes)) - mutual_weights1 @ (
+            mutual_weights2.T
+        )  # Redundancy
```

Comment:
I think this is equivalent to:

```suggestion
        r = 1 - (mutual_weights1 @ mutual_weights2.T).toarray()
```